docs/contributing/MODEL_MANAGER.md

```diff
@@ -0,0 +1,1214 @@
+taken to be the `models_dir` directory.
+
+`variant` is an enumerated string class with values `normal`,
+`inpaint` and `depth`. If needed, it can be imported if needed from
+either `invokeai.app.services.model_record_service` or
+`invokeai.backend.model_manager.config`.
+
+### ONNXSD2Config
+
+| **Field Name** | **Type**        |  **Description** |
+|----------------|-----------------|------------------|
+| `prediction_type`   | SchedulerPredictionType   | Scheduler prediction type to use, e.g. "epsilon" |
+| `upcast_attention`  | bool | Model requires its attention module to be upcast |
+
+The `SchedulerPredictionType` enum can be imported from either
+`invokeai.app.services.model_record_service` or
+`invokeai.backend.model_manager.config`.
+
+### Other config classes
+
+There are a series of such classes each discriminated by their
+`ModelFormat`, including `LoRAConfig`, `IPAdapterConfig`, and so
+forth. These are rarely needed outside the model manager's internal
+code, but available in `invokeai.backend.model_manager.config` if
+needed. There is also a Union of all ModelConfig classes, called
+`AnyModelConfig` that can be imported from the same file.
+
+### Limitations of the Data Model
+
+The config hierarchy has a major limitation in its handling of the
+base model type. Each model can only be compatible with one base
+model, which breaks down in the event of models that are compatible
+with two or more base models. For example, SD-1 VAEs also work with
+SD-2 models. A partial workaround is to use `BaseModelType.Any`, which
+indicates that the model is compatible with any of the base
+models. This works OK for some models, such as the IP Adapter image
+encoders, but is an all-or-nothing proposition.
+
+Another issue is that the config class hierarchy is paralleled to some
+extent by a `ModelBase` class hierarchy defined in
+`invokeai.backend.model_manager.models.base` and its subclasses. These
+are classes representing the models after they are loaded into RAM and
+include runtime information such as load status and bytes used. Some
+of the fields, including `name`, `model_type` and `base_model`, are
+shared between `ModelConfigBase` and `ModelBase`, and this is a
+potential source of confusion.
+
+** TO DO: ** The `ModelBase` code needs to be revised to reduce the
+duplication of similar classes and to support using the `key` as the
+primary model identifier.
+
+## Reading and Writing Model Configuration Records
+
+The `ModelRecordService` provides the ability to retrieve model
+configuration records from SQL or YAML databases, update them, and
+write them back.
+
+A application-wide `ModelRecordService` is created during API
+initialization and can be retrieved within an invocation from the
+`InvocationContext` object:
+
+```
+store = context.services.model_record_store
+```
+
+or from elsewhere in the code by accessing
+`ApiDependencies.invoker.services.model_record_store`.
+
+### Creating a `ModelRecordService` 
```

Comment:
I think its great that we are abstracting the underlying storage. I also think that the proposed class hierarchy and initialization logic could be simplified significantly via composition + dependency injection rather than inheritance. (I think it would also make it easier to write unit tests.)

I'm happy to write some pseudo code to explain what I mean if that would be helpful.

Reply:
I've not seen any examples of how to do dependency injection in Python. Some pseudo code would be appreciated.